src/multivariate/jointorderstatistics.jl

```diff
@@ -0,0 +1,156 @@
+    # issorted(ranks) && allunique(ranks)
+    return issorted(ranks; lt=_islesseq)
+end
+function _are_ranks_valid(ranks::AbstractRange, n)
+    first(ranks) ≥ 1 || return false
+    last(ranks) ≤ n || return false
+    return issorted(ranks) && allunique(ranks)
+end
+
+length(d::JointOrderStatistics) = length(d.ranks)
+function insupport(d::JointOrderStatistics, x::AbstractVector)
+    return length(d) == length(x) && issorted(x) && all(Base.Fix1(insupport, d.dist), x)
+end
+minimum(d::JointOrderStatistics) = Fill(minimum(d.dist), length(d))
+maximum(d::JointOrderStatistics) = Fill(maximum(d.dist), length(d))
+
+params(d::JointOrderStatistics) = tuple(params(d.dist)..., d.n, d.ranks)
+partype(d::JointOrderStatistics) = partype(d.dist)
+Base.eltype(::Type{<:JointOrderStatistics{D}}) where {D} = Base.eltype(D)
+Base.eltype(d::JointOrderStatistics) = eltype(d.dist)
+
+function logpdf(d::JointOrderStatistics, x::AbstractVector{<:Real})
+    n = d.n
+    ranks = d.ranks
+    lp = loglikelihood(d.dist, x)
+    T = typeof(lp)
+    lp += loggamma(T(n + 1))
+    length(ranks) == n && return lp
+    i = first(ranks)
+    xᵢ = first(x)
+    if i > 1  # _marginalize_range(d.dist, 0, i, -Inf, xᵢ, T)
+        lp += (i - 1) * logcdf(d.dist, xᵢ) - loggamma(T(i))
+    end
+    for (j, xⱼ) in Iterators.drop(zip(ranks, x), 1)
+        lp += _marginalize_range(d.dist, i, j, xᵢ, xⱼ, T)
+        i = j
+        xᵢ = xⱼ
+    end
+    if i < n  # _marginalize_range(d.dist, i, n + 1, xᵢ, Inf, T)
+        lp += (n - i) * logccdf(d.dist, xᵢ) - loggamma(T(n - i + 1))
+    end
+    return lp
+end
+
+# given ∏ₖf(xₖ), marginalize all xₖ for i < k < j
+function _marginalize_range(dist, i, j, xᵢ, xⱼ, T)
+    k = j - i - 1
+    k == 0 && return zero(T)
+    return k * T(logdiffcdf(dist, xⱼ, xᵢ)) - loggamma(T(k + 1))
+end
+
+function _rand!(rng::AbstractRNG, d::JointOrderStatistics, x::AbstractVector{<:Real})
+    n = d.n
+    if n == length(d.ranks)  # ranks == 1:n
+        # direct method, slower than inversion method for large `n` and distributions with
+        # fast quantile function or that use inversion sampling
+        rand!(rng, d.dist, x)
+        sort!(x)
+    else
+        # use exponential generation method with inversion, where for gaps in the ranks, we
+        # use the fact that the sum Y of k IID variables xₘ ~ Exp(1) is Y ~ Gamma(k, 1).
+        # Lurie, D., and H. O. Hartley. "Machine-generation of order statistics for Monte
+        # Carlo computations." The American Statistician 26.1 (1972): 26-27.
+        # this is slow if length(d.ranks) is close to n and quantile for d.dist is expensive,
+        # but this branch is probably taken when length(d.ranks) is small or much smaller than n.
+        T = typeof(one(eltype(x)))
+        s = zero(eltype(x))
+        i = 0
+        for (m, j) in zip(eachindex(x), d.ranks)
+            k = j - i
+            if k > 1
+                s += T(rand(rng, GammaMTSampler(Gamma{T}(T(k), T(1)))))
```

Comment:
Is this the method that would be picked by calling `rand(Gamma(...))` directly?

Reply:
It is, but only after checking the shape `k`, which we've already done:
https://github.com/JuliaStats/Distributions.jl/blob/ef42afb215cb08304dd64e309da8002c333caef0/src/univariate/continuous/gamma.jl#L103-L113
Plus, I benchmarked all samplers that apply for `Gamma(k, 1)` with `k>1`, and this one is the fastest.

Reply:
Maybe add a short comment in the code?